# Collaborative Filtering - Movie Recommendation Project
---

## BUSINESS PROBLEM
---
The online movie streaming platform (for example, superperfectmoviesblabla.tv) wants to develop a recommendation system using collaborative filtering. They aim to generate recommendations that encompass the opinions of the company community who have experimented with content-based recommendation systems. When users like a movie, they want to be recommended other movies that have similar likability patterns.

## DATASET STORY
---

The dataset has been provided by MovieLens, containing movies and the ratings given to these movies. The dataset includes approximately 2,000,000 ratings for around 27,000 movies.

## FEATURES
---
## movie.csv
- **movieId:** Unique movie number (UniqueID)
- **title:** Movie title

## rating.csv
- **userId:** Unique user number (UniqueID)
- **movieId:** Unique movie number (UniqueID)
- **rating:** Rating given to the movie by the user
- **timestamp:** Rating date

## STEPS
---
### Step 1: Data Preprocessing
### Step 2: Creating User-Movie DataFrame
### Step 3: Making Item-Based Movie Recommendations
### Step 4: Making User-Based Movie Recommendations
### Step 5: Preparation of Working Script
---

### *Libraries and Settings*
---

In [1]:
############################################################################################################################################
import pandas as pd
import numpy as np
import warnings
############################################################################################################################################
pd.set_option("display.max_columns", 500)
warnings.filterwarnings("ignore")
############################################################################################################################################

---
# Data Preprocessing
---

In [2]:
############################################################################################################################################
movies = pd.read_csv("/kaggle/input/movielens-25m-latest-dataset/movies.csv")
ratings = pd.read_csv("/kaggle/input/movielens-25m-latest-dataset/ratings.csv")
############################################################################################################################################
df = ratings.merge(movies, how="left")
df.head()
############################################################################################################################################

userId  movieId  rating   timestamp  \
0       1      296     5.0  1147880044   
1       1      306     3.5  1147868817   
2       1      307     5.0  1147868828   
3       1      665     5.0  1147878820   
4       1      899     3.5  1147868510   

                                              title  \
0                               Pulp Fiction (1994)   
1  Three Colors: Red (Trois couleurs: Rouge) (1994)   
2  Three Colors: Blue (Trois couleurs: Bleu) (1993)   
3                                Underground (1995)   
4                        Singin' in the Rain (1952)   

                        genres  
0  Comedy|Crime|Drama|Thriller  
1                        Drama  
2                        Drama  
3             Comedy|Drama|War  
4       Comedy|Musical|Romance

In [3]:
df.shape

(25000095, 6)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000095 entries, 0 to 25000094
Data columns (total 6 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
 4   title      object 
 5   genres     object 
dtypes: float64(1), int64(3), object(2)
memory usage: 1.1+ GB


In [5]:
# convert timestamp to datetime dtype
df["timestamp"] = pd.to_datetime(df["timestamp"], unit= "s").dt.date
df.head()

userId  movieId  rating   timestamp  \
0       1      296     5.0  2006-05-17   
1       1      306     3.5  2006-05-17   
2       1      307     5.0  2006-05-17   
3       1      665     5.0  2006-05-17   
4       1      899     3.5  2006-05-17   

                                              title  \
0                               Pulp Fiction (1994)   
1  Three Colors: Red (Trois couleurs: Rouge) (1994)   
2  Three Colors: Blue (Trois couleurs: Bleu) (1993)   
3                                Underground (1995)   
4                        Singin' in the Rain (1952)   

                        genres  
0  Comedy|Crime|Drama|Thriller  
1                        Drama  
2                        Drama  
3             Comedy|Drama|War  
4       Comedy|Musical|Romance

In [6]:
# check for missing values
df.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
title        0
genres       0
dtype: int64

It is important to use memory efficiently while working with matrix structures in recommendation systems which is required a 'User-Movies Matrix' for correlation calculation in the next step.

That's why we define a threshold to eliminate the movies rated at least 1000 times by users and the users rated at least 30 movies, which we think that might not reflect a collaboration with other users and their preferences.

In [7]:
############################################################################################################################################
# count of movies rated by users based on threshold of 1000
title_count = pd.DataFrame(df["movieId"].value_counts(), columns=["count"])
m_less_than_1000 = len(title_count[title_count["count"] < 1000])
m_more_than_1000 = len(title_count[title_count["count"] >= 1000])

print("*"*100)
print("Total Movie Count: ", len(title_count))
print("*"*100)
print("Movie count rated less than 1000 times: ", m_less_than_1000,"(", round((m_less_than_1000 / len(title_count))*100,2), " %)")
print("Movie count rated more than 1000 times: ", m_more_than_1000,"(", round((m_more_than_1000 / len(title_count))*100,2), " %)")
print("*"*100)

############################################################################################################################################
# count of users rated movies based on threshold of 30
user_count = pd.DataFrame(df["userId"].value_counts(), columns=["count"])
u_less_than_30 = len(user_count[user_count["count"] < 30])
u_more_than_30 = len(user_count[user_count["count"] >= 30])

print("*"*100)
print("Total User Count: ", len(user_count))
print("*"*100)
print("User count rated less than 30 times: ", u_less_than_30,"(", round((u_less_than_30 / len(user_count))*100,2), " %)")
print("User count rated more than 30 times: ", u_more_than_30,"(", round((u_more_than_30 / len(user_count))*100,2), " %)")
print("*"*100)
############################################################################################################################################

****************************************************************************************************
Total Movie Count:  59047
****************************************************************************************************
Movie count rated less than 1000 times:  55253 ( 93.57  %)
Movie count rated more than 1000 times:  3794 ( 6.43  %)
****************************************************************************************************
****************************************************************************************************
Total User Count:  162541
****************************************************************************************************
User count rated less than 30 times:  28650 ( 17.63  %)
User count rated more than 30 times:  133891 ( 82.37  %)
****************************************************************************************************


93.56 % of movies are rated less than 1000 and 17.63 users rated at least 30 movies.
We remove them to reduce matrix size for efficiency while keeping most of the information in dataset. 

In [8]:
movies_to_be_removed = title_count[title_count["count"] < 1000].index.to_list()
users_to_be_removed = user_count[user_count["count"] < 30].index.to_list()

rec_df = df[~df["movieId"].isin(movies_to_be_removed)]
rec_df = rec_df[~rec_df["userId"].isin(users_to_be_removed)]

display(rec_df)

userId  movieId  rating   timestamp  \
0              1      296     5.0  2006-05-17   
1              1      306     3.5  2006-05-17   
2              1      307     5.0  2006-05-17   
3              1      665     5.0  2006-05-17   
4              1      899     3.5  2006-05-17   
...          ...      ...     ...         ...   
25000090  162541    50872     4.5  2009-04-28   
25000091  162541    55768     2.5  2009-04-28   
25000092  162541    56176     2.0  2009-04-28   
25000093  162541    58559     4.0  2009-04-28   
25000094  162541    63876     5.0  2009-04-28   

                                                     title  \
0                                      Pulp Fiction (1994)   
1         Three Colors: Red (Trois couleurs: Rouge) (1994)   
2         Three Colors: Blue (Trois couleurs: Bleu) (1993)   
3                                       Underground (1995)   
4                               Singin' in the Rain (1952)   
...                                                    ...   
25000090                                Ratatouille (2007)   
25000091                                  Bee Movie (2007)   
25000092                    Alvin and the Chipmunks (2007)   
25000093                           Dark Knight, The (2008)   
25000094                                       Milk (2008)   

                               genres  
0         Comedy|Crime|Drama|Thriller  
1                               Drama  
2                               Drama  
3                    Comedy|Drama|War  
4              Comedy|Musical|Romance  
...                               ...  
25000090     Animation|Children|Drama  
25000091             Animation|Comedy  
25000092              Children|Comedy  
25000093      Action|Crime|Drama|IMAX  
25000094                        Drama  

[21483819 rows x 6 columns]

Now, 21.48 Million records are kept out of 25 Million although we reduced column count 93.56% and 17.63% of row count.

In [9]:
before = len(user_count) * len(title_count)
after = len(rec_df["userId"].value_counts()) * len(rec_df["movieId"].value_counts())

print("Matrix Size BEFORE reduction:", len(user_count), "*", len(title_count),"=", before)
print("Matrix Size AFTER reduction:", len(rec_df["userId"].value_counts()), "*", len(rec_df["movieId"].value_counts()),"=", after)
print("Matrix Size Reduction Rate:", round((before - after)/before, 2)*100, "%")
print("Information kept:", round(len(rec_df)/len(df), 2)*100, "%")

Matrix Size BEFORE reduction: 162541 * 59047 = 9597558427
Matrix Size AFTER reduction: 133891 * 3794 = 507982454
Matrix Size Reduction Rate: 95.0 %
Information kept: 86.0 %


In other words, we reduced matrix size by 95% while keeping 86% of information.

---
# Creating User-Movie DataFrame
---

Thus, we can create 'User-Movies Matrix' using pivot table.

In [10]:
user_movie_matrix = rec_df.pivot_table(index="userId", columns="movieId", values="rating")
user_movie_matrix.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           3.5     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           3.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      11      12      13      14      15      16      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  17      18      19      20      21      22      23      24      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     4.0     NaN     NaN     NaN     NaN     NaN   

movieId  25      26      27      28      29      30      31      32      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     4.5     NaN     NaN     4.5   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     NaN     NaN     NaN     5.0   

movieId  34      35      36      39      41      42      43      44      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     5.0     2.0     NaN     NaN     NaN     NaN   

movieId  45      46      47      48      50      52      57      58      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     5.0     NaN     5.0     NaN     NaN     NaN   

movieId  60      61      62      63      64      65      66      68      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     0.5     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  69      70      71      72      73      74      76      78      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN   

Let's select a movieId and get first 10 movies correlated with that movie which are similarly rated movieIds with selected movieId. 

In [11]:
# let's take movie with Id 8 as example
selected_movieId = user_movie_matrix.loc[:,8]
correlated_movieIds = user_movie_matrix.corrwith(selected_movieId).sort_values(ascending = False)[:10]
correlated_movieIds

movieId
8         1.000000
90522     0.974316
189203    0.965535
89118     0.919393
69524     0.869783
74228     0.859885
112804    0.856975
162082    0.852198
52694     0.848627
95105     0.834364
dtype: float64

---
# Making Item-Based Movie Recommendations
---

First, we define some function to make searching easier for us.

In [12]:
def find_movie_title_using_id(movieId):
    return rec_df[rec_df["movieId"] == movieId]["title"].values[0]

def find_movie_id_using_name(movie_name):
    return rec_df[rec_df["title"] == movie_name]["movieId"].values[0]

def find_movie_names_with_keyword(seached_movie_name):
    movie_names = list()
    [movie_names.append(name) for name in rec_df["title"] if seached_movie_name.lower() in name.lower() and name not in movie_names]
    return movie_names

Now, we seach for a movie using a key.

In [13]:
seached_key_word = "legend"
movie_full_name = find_movie_names_with_keyword(seached_key_word)
movie_full_name

['Legends of the Fall (1994)',
 'Anchorman: The Legend of Ron Burgundy (2004)',
 'Legend of Zorro, The (2005)',
 'I Am Legend (2007)',
 'Anchorman 2: The Legend Continues (2013)',
 "City Slickers II: The Legend of Curly's Gold (1994)",
 'Urban Legend (1998)',
 'Legend of Drunken Master, The (Jui kuen II) (1994)',
 'Urban Legends: Final Cut (2000)',
 'Legend (1985)',
 'Legend of Bagger Vance, The (2000)',
 "Legend of the Guardians: The Owls of Ga'Hoole (2010)"]

As we have found the full name of the movie (assuming that we seached for 'I Am Legend (2007)'), we get related movieId using defined function.

In [14]:
searched_movie_id = find_movie_id_using_name(movie_full_name[3])
searched_movie_id

56174

After getting movieIds correlated with selected movie, we can find first 10 recommendations.

In [15]:
selected_movieId = user_movie_matrix.loc[:,searched_movie_id]
correlated_movies = user_movie_matrix.corrwith(selected_movieId).sort_values(ascending = False)[1:11] # [1:11] defined as movie itself is in index 0
correlated_movieIds = correlated_movies.index.to_list()
correlated_rate = correlated_movies.to_list()

print("*"*100)
print("'", movie_full_name[3], "' - Recommendation List")
print("*"*100)
for no, movieId, rate in zip(range(1,11),correlated_movieIds, correlated_rate):
    movie_name = find_movie_title_using_id(movieId)
    print(no,"-", movie_name, "-"*(70 - len(movie_name)-len(str(no))), " - ", round(rate*100,2),"%")
print("*"*100)

****************************************************************************************************
' I Am Legend (2007) ' - Recommendation List
****************************************************************************************************
1 - Widows' Peak (1994) --------------------------------------------------  -  68.45 %
2 - I, Robot (2004) ------------------------------------------------------  -  61.91 %
3 - Hancock (2008) -------------------------------------------------------  -  55.55 %
4 - World War Z (2013) ---------------------------------------------------  -  52.76 %
5 - Book of Eli, The (2010) ----------------------------------------------  -  52.02 %
6 - Preacher's Wife, The (1996) ------------------------------------------  -  51.1 %
7 - Shanghai Triad (Yao a yao yao dao waipo qiao) (1995) -----------------  -  49.21 %
8 - Focus (2015) ---------------------------------------------------------  -  48.48 %
9 - Day After Tomorrow, The (2004) -----------------------

---
# Making User-Based Movie Recommendations
---

The purpose of this part is to find the movies that is most probable liked by a selected user.

Let's first remember 'User-Movie Matrix' created previously.

In [16]:
user_movie_matrix.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           3.5     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           3.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      11      12      13      14      15      16      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  17      18      19      20      21      22      23      24      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     4.0     NaN     NaN     NaN     NaN     NaN   

movieId  25      26      27      28      29      30      31      32      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     4.5     NaN     NaN     4.5   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     NaN     NaN     NaN     5.0   

movieId  34      35      36      39      41      42      43      44      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     5.0     2.0     NaN     NaN     NaN     NaN   

movieId  45      46      47      48      50      52      57      58      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     5.0     NaN     5.0     NaN     NaN     NaN   

movieId  60      61      62      63      64      65      66      68      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     0.5     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  69      70      71      72      73      74      76      78      \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN   

We select a random user and find the movieIds that he/she rated.

In [36]:
selected_userId = user_movie_matrix.sample(1).index[0]

print("*"*100)
print("Selected UserId:", selected_userId)
print("*"*100)

umm_for_selected_userId = user_movie_matrix[user_movie_matrix.index == selected_userId] # 'umm' is used as abbreviation of 'user_movie_matrix'
umm_for_selected_userId

****************************************************************************************************
Selected UserId: 48735
****************************************************************************************************


movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
48735       NaN     3.0     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      11      12      13      14      15      16      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  17      18      19      20      21      22      23      24      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  25      26      27      28      29      30      31      32      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  34      35      36      39      41      42      43      44      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  45      46      47      48      50      52      57      58      \
userId                                                                    
48735       NaN     NaN     4.0     NaN     4.5     NaN     NaN     NaN   

movieId  60      61      62      63      64      65      66      68      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  69      70      71      72      73      74      76      78      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  79      81      82      85      86      88      89      92      \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  93      94      95      97      100     101     102     103     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  104     105     107     110     111     112     117     118     \
userId                                                                    
48735       NaN     NaN     NaN     4.0     NaN     NaN     NaN     NaN   

movieId  122     123     125     126     132     135     140     141     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  144     145     147     149     150     151     153     154     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  155     156     157     158     159     160     161     162     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  163     164     165     168     169     170     171     172     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  173     174     175     176     177     179     180     181     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  185     186     188     191     193     194     195     196     \
userId                                                                    
48735       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  198     199     203     204     205     207     208     213     \
userId                                                                    
48735   

In [37]:
bool_for_selected_userId = umm_for_selected_userId.apply(lambda col: col.notnull(), axis=1)
bool_for_selected_userId

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
48735     False    True   False   False   False   False   False   False   

movieId  9       10      11      12      13      14      15      16      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  17      18      19      20      21      22      23      24      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  25      26      27      28      29      30      31      32      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  34      35      36      39      41      42      43      44      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  45      46      47      48      50      52      57      58      \
userId                                                                    
48735     False   False    True   False    True   False   False   False   

movieId  60      61      62      63      64      65      66      68      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  69      70      71      72      73      74      76      78      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  79      81      82      85      86      88      89      92      \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  93      94      95      97      100     101     102     103     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  104     105     107     110     111     112     117     118     \
userId                                                                    
48735     False   False   False    True   False   False   False   False   

movieId  122     123     125     126     132     135     140     141     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  144     145     147     149     150     151     153     154     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  155     156     157     158     159     160     161     162     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  163     164     165     168     169     170     171     172     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  173     174     175     176     177     179     180     181     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  185     186     188     191     193     194     195     196     \
userId                                                                    
48735     False   False   False   False   False   False   False   False   

movieId  198     199     203     204     205     207     208     213     \
userId                                                                    
48735   

In [38]:
movieIds_ratedby_selected_userId = bool_for_selected_userId.iloc[0].loc[lambda movieId: movieId == True]
list_movieIds_ratedby_selected_userId = bool_for_selected_userId.iloc[0].loc[lambda movieId: movieId == True].index.to_list()

print("Movie count rated by UserId '", selected_userId,"' =", len(list_movieIds_ratedby_selected_userId))

Movie count rated by UserId ' 48735 ' = 131


Now, we find all users who rated all movies rated by selected user

In [39]:
allUserIds_and_movies_ratedby_selected_userId = user_movie_matrix.loc[:, list_movieIds_ratedby_selected_userId]
allUserIds_and_movies_ratedby_selected_userId # 180 columns, the same with selected userId 

movieId  2       47      50      110     293     296     303     318     \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     5.0     NaN     NaN   
2           NaN     NaN     NaN     5.0     NaN     NaN     NaN     5.0   
3           NaN     NaN     5.0     NaN     5.0     5.0     NaN     4.0   
4           NaN     NaN     NaN     NaN     NaN     4.0     NaN     NaN   
5           NaN     5.0     5.0     NaN     NaN     4.0     NaN     NaN   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162537      NaN     NaN     NaN     NaN     NaN     NaN     NaN     5.0   
162538      NaN     4.0     NaN     NaN     NaN     2.5     NaN     NaN   
162539      NaN     NaN     NaN     4.0     NaN     NaN     NaN     NaN   
162540      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
162541      NaN     4.5     5.0     NaN     4.0     3.5     NaN     4.0   

movieId  356     364     367     434     480     508     527     551     \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           4.5     4.5     NaN     NaN     2.0     NaN     5.0     NaN   
3           4.0     NaN     NaN     NaN     2.0     NaN     4.0     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     4.0     NaN     NaN     NaN     NaN     NaN     NaN   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162537      5.0     NaN     NaN     NaN     NaN     NaN     NaN     3.0   
162538      4.0     NaN     NaN     NaN     NaN     NaN     4.5     2.0   
162539      5.0     NaN     NaN     NaN     4.0     NaN     NaN     NaN   
162540      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
162541      NaN     NaN     NaN     NaN     NaN     NaN     4.5     NaN   

movieId  589     593     750     780     784     788     858     904     \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           4.0     NaN     NaN     NaN     NaN     NaN     3.5     NaN   
3           4.0     4.0     NaN     3.5     NaN     NaN     5.0     NaN   
4           4.0     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
5           NaN     5.0     NaN     4.0     2.0     3.0     NaN     NaN   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162537      NaN     5.0     NaN     NaN     NaN     NaN     NaN     NaN   
162538      NaN     4.5     NaN     NaN     NaN     NaN     NaN     NaN   
162539      NaN     NaN     5.0     4.0     NaN     NaN     NaN     NaN   
162540      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
162541      NaN     4.0     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  1036    1183    1198    1204    1210    1213    1219    1221    \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     4.0     NaN     5.0     NaN     NaN     NaN   
3           NaN     NaN     3.0     NaN     4.0     5.0     NaN     5.0   
4           4.0     NaN     4.0     NaN     3.0     NaN     NaN     NaN   
5           NaN     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162537      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
162538      NaN     2.5     NaN     NaN     NaN     NaN     4.5     NaN   
162539      NaN     5.0     5.0     4.0     4.0     NaN     NaN     NaN   
162540      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
162541      NaN     NaN     4.5     NaN     3.5     NaN     NaN     NaN   

movieId  1225    1258    1295    1485    1608    1610    1617    1645    \
userId               

In [50]:
perc_threshold_rated_same_movies = 0.7 # the threshold for users who rated at least 70% of the same movies rated by selected user
count_threshold_rated_same_movies = len(list_movieIds_ratedby_selected_userId) * perc_threshold_rated_same_movies
userIds_rated_atleast_X_perc_same_movies_with_selected_userId = allUserIds_and_movies_ratedby_selected_userId[allUserIds_and_movies_ratedby_selected_userId.notnull().sum(axis=1) > count_threshold_rated_same_movies]
userIds_rated_atleast_X_perc_same_movies_with_selected_userId

movieId  2       47      50      110     293     296     303     318     \
userId                                                                    
426         NaN     5.0     4.5     4.5     4.0     4.5     NaN     4.5   
541         4.5     5.0     5.0     4.5     5.0     5.0     NaN     5.0   
548         4.0     4.0     5.0     4.5     4.0     4.5     3.5     5.0   
847         NaN     4.5     4.5     4.0     4.0     4.0     NaN     5.0   
901         3.5     4.5     4.0     4.0     5.0     5.0     NaN     4.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162271      1.0     2.5     1.0     2.0     3.0     5.0     NaN     1.5   
162286      5.0     4.0     5.0     4.0     4.0     5.0     NaN     5.0   
162334      3.5     5.0     4.0     4.0     4.0     4.0     NaN     4.0   
162349      1.0     5.0     5.0     5.0     5.0     5.0     NaN     5.0   
162516      2.5     4.0     5.0     3.5     4.0     4.5     3.0     4.0   

movieId  356     364     367     434     480     508     527     551     \
userId                                                                    
426         4.0     NaN     1.0     NaN     NaN     3.5     4.5     NaN   
541         5.0     5.0     5.0     NaN     5.0     NaN     5.0     4.5   
548         4.0     5.0     2.5     3.0     4.0     NaN     5.0     NaN   
847         3.0     3.5     NaN     NaN     3.5     2.5     4.0     4.0   
901         4.0     4.5     3.0     NaN     4.0     3.5     4.5     4.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162271      2.5     3.0     NaN     NaN     3.0     3.5     4.0     2.5   
162286      5.0     5.0     4.0     3.0     4.0     NaN     5.0     NaN   
162334      4.0     3.5     3.5     NaN     4.5     NaN     NaN     4.0   
162349      5.0     5.0     5.0     NaN     5.0     NaN     5.0     NaN   
162516      4.5     4.0     3.5     3.0     4.5     4.0     4.0     3.0   

movieId  589     593     750     780     784     788     858     904     \
userId                                                                    
426         3.5     5.0     2.0     2.5     NaN     NaN     5.0     4.5   
541         5.0     5.0     4.5     NaN     NaN     NaN     5.0     4.5   
548         4.5     4.5     NaN     3.5     2.5     1.5     NaN     NaN   
847         3.0     3.5     5.0     2.0     NaN     NaN     4.5     4.5   
901         4.0     4.0     4.0     4.0     3.0     2.5     4.5     4.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162271      3.0     4.0     5.0     2.0     2.0     NaN     3.5     3.0   
162286      5.0     NaN     3.0     4.0     3.0     3.0     5.0     NaN   
162334      4.5     4.0     NaN     4.0     NaN     NaN     5.0     4.0   
162349      5.0     5.0     NaN     5.0     5.0     NaN     NaN     NaN   
162516      5.0     5.0     4.0     4.0     3.0     3.5     5.0     4.0   

movieId  1036    1183    1198    1204    1210    1213    1219    1221    \
userId                                                                    
426         4.5     NaN     4.0     4.0     2.5     3.5     3.5     5.0   
541         5.0     NaN     5.0     5.0     5.0     5.0     4.5     5.0   
548         5.0     NaN     4.5     NaN     4.0     5.0     NaN     NaN   
847         3.5     1.5     4.5     5.0     3.5     3.5     5.0     5.0   
901         4.0     NaN     4.5     5.0     3.5     4.5     4.0     4.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
162271      1.5     2.5     4.0     2.0     3.0     4.5     4.0     3.5   
162286      5.0     NaN     4.0     NaN     4.5     5.0     4.0     3.0   
162334      4.0     NaN     3.5     NaN     4.0     4.0     3.5     4.5   
162349      5.0     5.0     5.0     NaN     5.0     5.0     NaN     NaN   
162516      4.0     3.0     4.5     3.5     5.0     4.0     3.5     5.0   

movieId  1225    1258    1295    1485    1608    1610    1617    1645    \
userId               

In [46]:
corr_of_userIds_rated_same_movies_with_selected_userId = userIds_rated_atleast_X_perc_same_movies_with_selected_userId.T.corr().unstack()
corr_of_userIds_rated_same_movies_with_selected_userId

userId  userId
426     426       1.000000
        541       0.428736
        548       0.549978
        847       0.202527
        901       0.420778
                    ...   
162516  162271    0.185880
        162286    0.160912
        162334    0.191996
        162349    0.195723
        162516    1.000000
Length: 938961, dtype: float64

In [47]:
userIds_corr_df = pd.DataFrame(corr_of_userIds_rated_same_movies_with_selected_userId, columns=["corr"])
userIds_corr_df.index.names = ["userId_1","userId_2"]
userIds_corr_df.reset_index(inplace=True)
userIds_corr_df

userId_1  userId_2      corr
0            426       426  1.000000
1            426       541  0.428736
2            426       548  0.549978
3            426       847  0.202527
4            426       901  0.420778
...          ...       ...       ...
938956    162516    162271  0.185880
938957    162516    162286  0.160912
938958    162516    162334  0.191996
938959    162516    162349  0.195723
938960    162516    162516  1.000000

[938961 rows x 3 columns]

In [51]:
corr_threshold = 0.5 # users who have at least 60% correlation with selected user
final_users_corr_df = userIds_corr_df[(userIds_corr_df["userId_1"] == selected_userId) & (userIds_corr_df["corr"] >= corr_threshold)].sort_values(by="corr", ascending=False) # set threshold and sort correlations
final_users_corr_df

userId_1  userId_2      corr
291000     48735     48735  1.000000
291482     48735    132976  0.594367
291269     48735     95055  0.579613
291548     48735    143879  0.573067
291480     48735    132747  0.550089
291631     48735    155408  0.548828
290909     48735     34164  0.545303
290854     48735     26928  0.541808
291518     48735    139105  0.538599
291262     48735     94154  0.538116
291045     48735     57079  0.537037
291169     48735     78475  0.514038
291374     48735    112971  0.512629
291258     48735     93424  0.511699
291016     48735     50319  0.504943
290968     48735     42202  0.501148

In [62]:
list_users_to_filter = final_users_corr_df["userId_2"].to_list()

final_rec_df = user_movie_matrix.loc[list_users_to_filter,list_movieIds_ratedby_selected_userId]
final_rec_df

movieId  2       47      50      110     293     296     303     318     \
userId                                                                    
48735       3.0     4.0     4.5     4.0     4.5     4.5     2.5     4.5   
132976      2.5     4.5     3.5     5.0     4.0     4.5     3.5     4.0   
95055       3.0     4.0     4.0     3.5     3.0     4.0     NaN     4.5   
143879      2.0     5.0     3.5     4.5     3.5     4.0     2.0     5.0   
132747      3.0     4.5     3.5     2.0     4.0     5.0     NaN     4.5   
155408      3.5     5.0     5.0     5.0     4.0     5.0     NaN     5.0   
34164       3.5     5.0     5.0     4.0     5.0     4.5     NaN     5.0   
26928       NaN     4.5     4.5     4.5     4.5     4.5     NaN     5.0   
139105      1.5     4.5     5.0     3.5     4.5     5.0     NaN     5.0   
94154       3.5     3.5     3.5     4.0     4.0     5.0     NaN     4.0   
57079       NaN     4.5     4.0     4.0     4.0     4.5     NaN     5.0   
78475       NaN     4.0     3.5     3.5     3.5     4.0     NaN     4.0   
112971      3.0     3.0     4.5     3.5     3.5     3.5     NaN     4.5   
93424       3.0     4.5     5.0     4.0     4.0     5.0     3.5     4.5   
50319       3.0     5.0     4.5     4.5     4.5     5.0     NaN     5.0   
42202       4.0     NaN     5.0     4.5     5.0     4.5     NaN     4.5   

movieId  356     364     367     434     480     508     527     551     \
userId                                                                    
48735       5.0     5.0     4.5     3.5     3.0     4.5     5.0     4.0   
132976      4.0     4.0     1.5     3.0     3.5     3.5     5.0     NaN   
95055       4.0     4.0     2.5     NaN     3.5     4.5     4.5     3.5   
143879      3.5     4.5     4.0     1.0     3.5     NaN     5.0     4.0   
132747      4.0     3.0     4.0     2.0     3.0     NaN     4.0     NaN   
155408      5.0     3.0     3.5     NaN     3.0     3.5     5.0     NaN   
34164       5.0     NaN     3.5     NaN     4.0     4.0     5.0     NaN   
26928       4.5     NaN     NaN     NaN     4.0     4.0     5.0     NaN   
139105      5.0     3.5     2.0     NaN     3.5     4.0     5.0     3.5   
94154       3.5     4.5     2.5     NaN     3.5     NaN     4.5     3.0   
57079       4.5     4.0     NaN     NaN     4.0     4.5     5.0     NaN   
78475       4.0     2.5     NaN     NaN     3.5     NaN     4.0     NaN   
112971      4.0     NaN     3.0     NaN     2.5     3.5     4.0     NaN   
93424       4.0     4.5     3.0     NaN     4.0     4.0     4.5     4.5   
50319       5.0     4.5     2.5     2.0     5.0     4.5     5.0     3.5   
42202       5.0     4.5     NaN     NaN     4.0     NaN     4.5     4.0   

movieId  589     593     750     780     784     788     858     904     \
userId                                                                    
48735       3.0     4.5     4.0     3.0     3.5     3.0     5.0     4.5   
132976      NaN     3.5     NaN     NaN     NaN     NaN     5.0     NaN   
95055       3.0     4.0     3.5     3.0     2.0     2.0     4.5     NaN   
143879      5.0     4.0     NaN     2.0     2.0     2.0     4.0     NaN   
132747      1.5     3.5     NaN     3.0     NaN     2.0     4.5     NaN   
155408      1.0     4.5     NaN     3.0     3.0     2.0     5.0     4.0   
34164       5.0     5.0     4.0     3.0     NaN     NaN     5.0     NaN   
26928       4.0     4.5     NaN     4.0     3.0     NaN     5.0     NaN   
139105      4.0     5.0     4.5     4.0     NaN     1.5     5.0     NaN   
94154       4.0     4.5     2.5     3.5     2.5     NaN     4.5     4.0   
57079       3.5     4.5     NaN     3.5     NaN     3.0     NaN     NaN   
78475       2.5     3.5     4.0     3.0     NaN     2.5     4.5     3.5   
112971      3.0     3.5     4.5     NaN     NaN     NaN     5.0     NaN   
93424       3.5     5.0     4.5     3.0     2.5     2.5     5.0     4.5   
50319       4.5     5.0     4.0     4.0     NaN     2.5     5.0     4.0   
42202       4.5     5.0

---
# Preparation of Working Script
---

---
## I hope you found the analysis insightful and informative!

## Your feedback is greatly appreciated, and I welcome any suggestions for improvement. Feel free to reach out with any questions or comments. 👍🏼
---